Add LLVM CFI support to the Rust compiler

rcvalle opened the pull request:
This PR adds LLVM Control Flow Integrity (CFI) support to the Rust compiler. It initially provides forward-edge control flow protection for Rust-compiled code only by aggregating function pointers in groups identified by their number of arguments.

Forward-edge control flow protection for C or C++ and Rust -compiled code "mixed binaries" (i.e., for when C or C++ and Rust -compiled code share the same virtual address space) will be provided in later work as part of this project by defining and using compatible type identifiers (see Type metadata in the design document in the tracking issue #89653).

LLVM CFI can be enabled with -Zsanitizer=cfi and requires LTO (i.e., -Clto).

Thank you, @eddyb and @pcc, for all the help!

rcvalle committed:
Add LLVM CFI support to the Rust compiler

This commit adds LLVM Control Flow Integrity (CFI) support to the Rust
compiler. It initially provides forward-edge control flow protection for
Rust-compiled code only by aggregating function pointers in groups
identified by their number of arguments.

Forward-edge control flow protection for C or C++ and Rust -compiled
code "mixed binaries" (i.e., for when C or C++ and Rust -compiled code
share the same virtual address space) will be provided in later work as
part of this project by defining and using compatible type identifiers
(see Type metadata in the design document in the tracking issue #89653).

LLVM CFI can be enabled with -Zsanitizer=cfi and requires LTO (i.e.,
-Clto).

rust-highfive commented:
r? @estebank

(rust-highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
:warning: **Warning** :warning:

* These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3833132724) [(plain)](https://github.com/rust-lang/rust/commit/fb5ac379fa3e6ccf193c81d0794cfadd1394b97b/checks/3833132724/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0412]: cannot find type `Value` in this scope
   --> src/builder.rs:918:49
    |
918 |     fn type_metadata(&mut self, _function: &'ll Value, _typeid: String) {
    |
    |
   ::: /cargo/git/checkouts/gccjit.rs-13c2e290f2fb9e4d/2d4fea7/src/lvalue.rs:43:1
    |
43  | pub struct LValue<'ctx> {
    | ----------------------- similarly named struct `LValue` defined here
    |
help: there is an enum variant `rustc_ast::CaptureBy::Value` and 4 others; try using the variant's enum
    |
918 |     fn type_metadata(&mut self, _function: &'ll rustc_ast::CaptureBy, _typeid: String) {
    |                                                 ~~~~~~~~~~~~~~~~~~~~
918 |     fn type_metadata(&mut self, _function: &'ll rustc_ast::SelfKind, _typeid: String) {
    |                                                 ~~~~~~~~~~~~~~~~~~~
918 |     fn type_metadata(&mut self, _function: &'ll rustc_hir::CaptureBy, _typeid: String) {
    |                                                 ~~~~~~~~~~~~~~~~~~~~
918 |     fn type_metadata(&mut self, _function: &'ll rustc_middle::ty::ConstKind, _typeid: String) {
help: a struct with a similar name exists
    |
    |
918 |     fn type_metadata(&mut self, _function: &'ll LValue, _typeid: String) {


error[E0261]: use of undeclared lifetime name `'ll`
    |
    |
918 |     fn type_metadata(&mut self, _function: &'ll Value, _typeid: String) {
    |                                             ^^^ undeclared lifetime
    |
    = help: if you want to experiment with in-band lifetime bindings, add `#![feature(in_band_lifetimes)]` to the crate attributes
help: consider introducing lifetime `'ll` here
    |
385 | impl<'ll, 'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
    |      ++++
help: consider introducing lifetime `'ll` here
    |
918 |     fn type_metadata<'ll>(&mut self, _function: &'ll Value, _typeid: String) {

error[E0046]: not all trait items implemented, missing: `type_test`
  --> src/intrinsic/mod.rs:76:1
   |
   |
76 | impl<'a, 'gcc, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'a, 'gcc, 'tcx> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `type_test` in implementation
   |
   = help: implement the missing item: `fn type_test(&mut self, _: <Self as BackendTypes>::Value, _: <Self as BackendTypes>::Value) -> <Self as BackendTypes>::Value { todo!() }`
error[E0308]: mismatched types
   --> src/builder.rs:922:55
    |
    |
922 |     fn typeid_metadata(&mut self, _typeid: String) -> Self::Value {
    |        ---------------                                ^^^^^^^^^^^ expected struct `RValue`, found `()`
    |        |
    |        implicitly returns `()` as its body has no tail or `return` expression
Some errors have detailed explanations: E0046, E0261, E0308, E0412.
For more information about an error, try `rustc --explain E0046`.
error: could not compile `rustc_codegen_gcc` due to 4 previous errors
Build completed unsuccessfully in 0:03:19
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3833187164) [(plain)](https://github.com/rust-lang/rust/commit/7c25df1d875e8dd2fba370b7f34bdd11cca334b4/checks/3833187164/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0046]: not all trait items implemented, missing: `type_test`
  --> src/intrinsic/mod.rs:76:1
   |
76 | impl<'a, 'gcc, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'a, 'gcc, 'tcx> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `type_test` in implementation
   |
   = help: implement the missing item: `fn type_test(&mut self, _: <Self as BackendTypes>::Value, _: <Self as BackendTypes>::Value) -> <Self as BackendTypes>::Value { todo!() }`
error[E0308]: mismatched types
   --> src/builder.rs:922:55
    |
    |
922 |     fn typeid_metadata(&mut self, _typeid: String) -> Self::Value {
    |        ---------------                                ^^^^^^^^^^^ expected struct `RValue`, found `()`
    |        |
    |        implicitly returns `()` as its body has no tail or `return` expression
Some errors have detailed explanations: E0046, E0308.
For more information about an error, try `rustc --explain E0046`.
error: could not compile `rustc_codegen_gcc` due to 2 previous errors
Build completed unsuccessfully in 0:03:40
```

</details>

nagisa commented:
Huh, there seem to be more caveats with `CFI` compared to other sanitizers as far as I can tell at a glance. I would like to see this be landed behind a `-Zunstable-options` or somesuch initially to give us more leeway in adjusting the specifics of the implementation and interactions with other ecosystems.

rcvalle committed:
Add documentation for LLVM CFI support

This commit adds initial documentation for LLVM Control Flow Integrity
(CFI) support to the Rust compiler (see #89652 and #89653).

nagisa commented:
This does look pretty reasonable first step in the implementation.

@bors r+

bors commented:
:pushpin: Commit d9e56b374040e01f64ded568bc61afe22f128d9a has been approved by `nagisa`

<!-- @bors r=nagisa d9e56b374040e01f64ded568bc61afe22f128d9a -->
<!-- homu: {"type":"Approved","sha":"d9e56b374040e01f64ded568bc61afe22f128d9a","approver":"nagisa"} -->

rcvalle commented:
Thank you for your time, @nagisa!
